src/libtiled/tilelayer.cpp

```diff
@@ -360,16 +372,34 @@ void TileLayer::rotateHexagonal(RotateDirection direction)
             dest.setFlippedAntiDiagonally((mask & 2) != 0);
             dest.setRotatedHexagonal120((mask & 1) != 0);
 
-            if (direction == RotateRight)
-                newGrid[x * newWidth + (mHeight - y - 1)] = dest;
-            else
-                newGrid[(mWidth - x - 1) * newWidth + y] = dest;
+            Hex rotatedHex(x, y, staggerIndex, staggerAxis);
+            rotatedHex -= center;
+            rotatedHex.rotate(direction);
+            rotatedHex += newCenter;
+
+            QPoint rotatedPoint = rotatedHex.toStaggered(staggerIndex, staggerAxis);
+
+            int index = rotatedPoint.y() * newWidth + rotatedPoint.x();
+
+            newGrid[index] = dest;
         }
     }
 
     mWidth = newWidth;
     mHeight = newHeight;
     mGrid = newGrid;
+
+    QRect filledRect = region().boundingRect();
+
+    if (staggerAxis == Map::StaggerY) {
+        if (filledRect.y() & 1)
+            filledRect.adjust(0, -1, 0, 0);
```

Comment:
This was admittedly a bit of a lazy solution. Though I believe we don't have the inherent ability to change the stagger index from a TileLayer. The solution could be to omit the staggerIndex and staggerAxis parameters, and in place pass the map reference. I was thinking it would be preferable to keep them separate,  but I'll go ahead and implement that way if you think I'd be better.